src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs

```diff
@@ -24,70 +25,67 @@ public sealed partial class JsonSourceGenerator : IIncrementalGenerator
         public void Initialize(IncrementalGeneratorInitializationContext context)
         {
 #if LAUNCH_DEBUGGER
-            Diagnostics.Debugger.Launch();
+            System.Diagnostics.Debugger.Launch();
 #endif
-            IncrementalValuesProvider<ClassDeclarationSyntax> classDeclarations = context.SyntaxProvider
+            IncrementalValueProvider<KnownTypeSymbols> knownTypeSymbols = context.CompilationProvider
+                .Select((compilation, _) => new KnownTypeSymbols(compilation));
+
+            IncrementalValuesProvider<(ContextGenerationSpec?, ImmutableEquatableArray<DiagnosticInfo>)> contextGenerationSpecs = context.SyntaxProvider
                 .ForAttributeWithMetadataName(
 #if !ROSLYN4_4_OR_GREATER
                     context,
 #endif
                     Parser.JsonSerializableAttributeFullName,
                     (node, _) => node is ClassDeclarationSyntax,
-                    (context, _) => (ClassDeclarationSyntax)context.TargetNode);
-
-            IncrementalValueProvider<SourceGenerationSpec?> sourceGenSpec = context.CompilationProvider
-                .Combine(classDeclarations.Collect())
+                    (context, _) => (ContextClass: (ClassDeclarationSyntax)context.TargetNode, context.SemanticModel))
```

Comment:
Aren't these syntax nodes going to be invalidated on every incremental compilation?  I thought the intent was to do the extraction / parsing here and feed to the next step of the pipeline just the extracted data?

Reply:
As written, won't it re-execute the Select and thus the full parse for every attribute when only one of them changes? Is there nothing that can be extracted from the node here that's less expensive than the full parse?

Reply:
That's right. The goal of this change is to avoid triggering source production for every context class in cases where the parser produces different outputs for a subset of the declared context classes.